vault/acl_test.go

```diff
@@ -415,6 +415,36 @@ func TestACL_AllowOperation(t *testing.T) {
 	})
 }
 
+func BenchmarkACLAllowOperation(b *testing.B) {
+	b.ReportAllocs()
+	ns := namespace.RootNamespace
+	policy, err := ParseACLPolicy(ns, `
+name = "dev"
+path "foo/bar" {
+	policy = "write"
+}
+`)
+
+	if err != nil {
+		b.Fatalf("err: %v", err)
+	}
+	ctx := namespace.ContextWithNamespace(context.Background(), ns)
+	acl, err := NewACL(ctx, []*Policy{policy})
+	if err != nil {
+		b.Fatalf("err: %v", err)
+	}
+
+	request := &logical.Request{
+		Path:      "/foo/bar",
+		Data:      make(map[string]interface{}),
+		Operation: logical.ListOperation,
+	}
+
+	for i := 0; i < b.N; i++ {
+		acl.AllowOperation(ctx, request, true)
```

Comment:
can you store the value here? i wonder if that brings us up to the same number of allocations?

Reply:
Note that I added two benchmarks, and the one I cited above is the other one, BenchmarkHasAccess.  That one invokes AllowOperation via hasAccess, which results in the return value being assigned.  This one (BenchmarkACLAllowOperation) was added to show the extra overhead hasAccess gives on top of AllowOperation, and to demonstrate that the allocation occurs in AllowOperation.

Benchmark results with AllowOperation returning a pointer:
```
BenchmarkACLAllowOperation-8   	10000000	       106 ns/op	      48 B/op	       1 allocs/op
BenchmarkHasAccess-8   	10000000	       127 ns/op	      48 B/op	       1 allocs/op
```

Returning a struct:
```
BenchmarkACLAllowOperation-8   	30000000	        49.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkHasAccess-8   	20000000	        74.3 ns/op	       0 B/op	       0 allocs/op
```

Changing BenchmarkACLAllowOperation to have a body like this had no impact on allocations, whether ret was ACLResults or *ACLResults:
```
	var ret ACLResults
	for i := 0; i < b.N; i++ {
		ret = acl.AllowOperation(ctx, request, true)
	}
	b.Log(ret)
```